x-pack/plugin/ml/src/main/java/org/elasticsearch/xpack/ml/utils/ExceptionCollectionHandling.java

```diff
@@ -0,0 +1,66 @@
+/*
+ * Copyright Elasticsearch B.V. and/or licensed to Elasticsearch B.V. under one
+ * or more contributor license agreements. Licensed under the Elastic License
+ * 2.0; you may not use this file except in compliance with the Elastic License
+ * 2.0.
+ *
+ * This file was contributed to by generative AI
+ */
+
+package org.elasticsearch.xpack.ml.utils;
+
+import org.elasticsearch.ElasticsearchStatusException;
+import org.elasticsearch.common.util.concurrent.AtomicArray;
+import org.elasticsearch.rest.RestStatus;
+
+import java.util.List;
+
+public class ExceptionCollectionHandling {
+
+    public static ElasticsearchStatusException exceptionArrayToStatusException(AtomicArray<Exception> failures, String message) {
+
+        List<Exception> caughtExceptions = failures.asList();
+        if (caughtExceptions.isEmpty()) {
+            throw new ElasticsearchStatusException("No exceptions caught", RestStatus.INTERNAL_SERVER_ERROR);
+        } else {
+
+            boolean allElasticsearchStatusException = true;
+            boolean allElasticsearchStatusException4xx = true;
+            boolean allSameCode = true;
+            int firstCode = -1;
+
+            for (Exception exception : caughtExceptions) {
+                if (exception instanceof ElasticsearchStatusException == false) {
+                    allElasticsearchStatusException = false;
+                    break;
+                }
+                ElasticsearchStatusException elasticsearchStatusException = (ElasticsearchStatusException) exception;
+                if (firstCode == -1) {
+                    firstCode = elasticsearchStatusException.status().getStatus();
+                } else if (false == (firstCode == elasticsearchStatusException.status().getStatus())) {
+                    allSameCode = false;
+                }
+
+                if (elasticsearchStatusException.status().getStatus() < 400 || elasticsearchStatusException.status().getStatus() >= 500) {
+                    allElasticsearchStatusException4xx = false;
+                }
+            }
+
+            if (allElasticsearchStatusException && allElasticsearchStatusException4xx) {
+                if (allSameCode) {
+                    return new ElasticsearchStatusException(
+                        message,
+                        ((ElasticsearchStatusException) caughtExceptions.get(0)).status(),
+                        caughtExceptions.get(0)
+                    );
+                } else {
+                    return new ElasticsearchStatusException(message, RestStatus.REQUEST_TIMEOUT);
```

Comment:
I think for this case it's better to return a code that actually happened rather than always return timeout.

Reply:
I removed the branch for having the same status. Now, this function returns the first exception if they are all 4XX statuses, and 500 status if any of the statuses are outside of the 4xx range or are not elasticsearchExceptions.